dotcom-rendering/src/server/index.article.web.ts

```diff
@@ -30,7 +32,10 @@ const enhanceBadge = (badge?: FEArticleBadgeType) =>
 		: undefined;
 
 export const enhanceArticleType = (body: unknown): FEArticleType => {
-	const data = validateAsArticleType(body);
+	const validated = validateAsArticleType(body);
+	// addImageIDs needs to take account of both main media elements
+	// and block elements, so it needs to be executed here
+	const data = addImageIDs(validated);
```

Comment:
It could but I've done it like this because this function is basically just another enhancer and all the other enhancers follow the same signature. You pass the old data in and get new, enhanced, data back. It's true that the other enhancers are taking in / passing back `blocks` but it's the same idea.

It's because of this pattern than you can do things like

```
		return new Enhancer(elements)
			.addLightboxData()
			.stripCaptions()
			.removeCredit()
			.addMultiImageElements()
			.addTitles()
			.addCaptionsToMultis()
			.addCaptionsToImages().elements;
```

Reply:
> this function is basically just another enhancer and all the other enhancers follow the same signature. You pass the old data in and get new, enhanced, data back.

We neigher have a `Enhancer<T>` signature nor is this function following the enhancer pattern, which implements a Class that mutates its array internally. This could be very confusing for future developers as there’s currently nothing in this code that ties these two concepts together.

> It's because of this pattern than you can do things like [example]

As I understand it, we’re not using that pattern here, so this should not be relevant. Maybe we should make this an actual enhancer?

--- 

Worth noting that a purely functional approach may also yield similar results, e.g.

```ts
const addPosition = (elements: FEElement[], offset = 0): FEElement[] => {
	let position = offset; 
	const withPosition: FEElement[] = elements.map((element) => {
		if (
			thisElement._type ===
			'model.dotcomrendering.pageElements.ImageBlockElement'
		) {
			position += 1;
			return {
				...element,
				position,
			}
		} else {
			return element;
		}
	});	
	return { withPosition, positionOffset: position };
};
```

Reply:
I've [updated](https://github.com/guardian/dotcom-rendering/pull/7129/commits/8dfc2e31b9fe6f00f0c2051793a3dea1bbf08c1b) the `addImageIds` function to only return the data it affects 😄